python/xir/ir.lark

```diff
@@ -4,11 +4,14 @@ include: "use" name ";"
 
 // the main circuit
 circuit: statement+
-statement: gatestmt | gate_def | operator_def | script_options
+statement: gatestmt | adjoint_gatestmt | ctrl_gatestmt | gate_def | operator_def | script_options
 gatestmt: name (params | options_dict) "|" wires ";"
 opstmt: expr "," obs_group";"
 script_options: _OPTIONS ":" (option ";")+ _END ";"
 
+adjoint_gatestmt: _ADJOINT gatestmt
+ctrl_gatestmt: _CTRL wires gatestmt
```

Comment:
Does this mean the following syntax is invalid?
```
adjoint ctrl[0] H | [1];
```
Ideally, I think we should be able to compose the two; however, we would need to be careful so that specifying `adjoint` (or `ctrl`) more than once is illegal.

Reply:
Yes, it currently is. Good point though. Could perhaps include that case as well. 🤔

Reply:
> we would need to be careful so that specifying adjoint (or ctrl) more than once is illegal.

Is it though? The adjoint of an adjoint, or a control of an adjoint of a control, is perfectly well defined.

Or is it more from the perspective of clean design?

Reply:
> The adjoint of an adjoint, or a control of an adjoint of a control, is perfectly well defined.

Ah, of course! Does PennyLane support syntax which directly translates to taking an adjoint of an adjoint (or the control of an adjoint of a control)? If yes, then we should support multiple `adjoint` (and `ctrl`) modifiers in gate application statements. If not, then we have the freedom to do whichever of the following is easier to implement:
1. Restrict `adjoint` and `ctrl` modifiers such that only one of each can be applied.
2. Enable arbitrary sequences of `adjoint` and `ctrl` modifiers.

> Or is it more from the perspective of clean design?

Thanks for giving me the benefit of the doubt, but you had it right the first time 🙂.

Reply:
>Does PennyLane support syntax which directly translates to taking an adjoint of an adjoint (or the control of an adjoint of a control)? If yes, then we should support multiple adjoint (and ctrl) modifiers in gate application statements

Yep! Although I suppose we could always have PennyLane _simplify_ to a single adjoint and a single ctrl (if possible).

Reply:
+1 for the ability to use adj and ctrl together.

I still have a soft spot for having control be `H | [1] | [0];` rather than `ctrl[0] H | [1];`. Can someone persuade me why that's a bad idea :laughing: One drawback would be only specifying control once.

Reply:
It's now possible to use any sequence of `ctrl` and `adjoint`, and it should work correctly.

For example,
```
adjoint adjoint ctrl[0, 1] adjoint ry(2.4) | [2];
```
would result with `adjoint=True` and `ctrl_wires=(0, 1)`.

Reply:
> I still have a soft spot for having control be `H | [1] | [0];` rather than `ctrl[0] H | [1];`. Can someone persuade me why that's a bad idea 😆 One drawback would be only specifying control once.

@trbromley, as someone with relatively little experience in PennyLane or similar, I think  `H | [1] | [0];` is pretty confusing.  Sure, this may look clean once you understand the grammar, but I think the readability advantage of `ctrl[...]` is more important.  The other advantage of `ctrl` is that it resembles the syntax of [OpenQASM 3](https://arxiv.org/pdf/2104.14722.pdf).


> It's now possible to use any sequence of `ctrl` and `adjoint`, and it should work correctly.
> 
> For example,
> 
> ```
> adjoint adjoint ctrl[0, 1] adjoint ry(2.4) | [2];
> ```
> 
> would result with `adjoint=True` and `ctrl_wires=(0, 1)`.

Awesome!  Just to confirm, we only really care about the parity of the number of times `adjoint` is specified, right?  In other words, can we reduce any statement with a sequence of modifiers to
```java
ctrl [ /*Union of all wires in ctrl modifiers*/ ] gate_name (...) | [...];
```
if `adjoint` is specified an even number of times and 
```java
ctrl [ /*Union of all wires in ctrl modifiers*/ ] adjoint gate_name (...) | [...];
```
if `adjoint` is specified an odd number of times, right?

Reply:
Yes, that's basically exactly how this works now.